File: wqr/quantile_density.py
"""
Nonparametric Wavelet Quantile Density Estimation.

Implements the Daubechies-Lagarias wavelet-based estimator for the
quantile density function q(p) = 1/f(F^{-1}(p)), with:
  - Linear wavelet estimate
  - Hard thresholding
  - Local linear smoothing

Converted from the MATLAB code by Chesneau, Dewan & Doosti.
"""

import numpy as np
from dataclasses import dataclass
from typing import Optional

import pywt

from .utils import safe_print


@dataclass
class QuantileDensityResult:
    """Container for quantile density estimation results."""
    grid: np.ndarray                # probability grid p in (0,1)
    true_qd: Optional[np.ndarray]   # true quantile density (if known)
    linear_estimate: np.ndarray     # linear wavelet estimate
    thresholded_estimate: np.ndarray  # after hard thresholding
    smoothed_estimate: np.ndarray   # after local linear smoothing
    sample_size: int
    j0: int
    bandwidth: float

    def summary(self):
        lines = [
            "",
            "╔══════════════════════════════════════════════════════╗",
            "║  Wavelet Quantile Density Estimation — Summary      ║",
            "╚══════════════════════════════════════════════════════╝",
            "",
            f"  Sample size  : {self.sample_size}",
            f"  Coarsest j₀  : {self.j0}",
            f"  Bandwidth    : {self.bandwidth:.4f}",
            f"  Grid points  : {len(self.grid)}",
            "",
        ]
        if self.true_qd is not None:
            ise_lin = np.mean((self.true_qd - self.linear_estimate) ** 2)
            ise_thr = np.mean((self.true_qd - self.thresholded_estimate) ** 2)
            ise_smo = np.mean((self.true_qd - self.smoothed_estimate) ** 2)
            lines += [
                "  ISE (Linear)      : {:.6f}".format(ise_lin),
                "  ISE (Thresholded) : {:.6f}".format(ise_thr),
                "  ISE (Smoothed)    : {:.6f}".format(ise_smo),
                "",
            ]
        safe_print("\n".join(lines))
        return self


def _simpson_coef(j0, k, y):
    """
    Estimate wavelet scaling coefficient ĉ_{j0,k} using Simpson's rule.
    Mirrors the MATLAB ``simp.m``.
    """
    n = len(y)
    sorted_y = np.sort(y)

    # Scaling function evaluated at 2^{j0}*F_n(y_i) - k
    scale = 2 ** (j0 / 2)
    coefficients = np.zeros(n)

    for i in range(n):
        fn_yi = (i + 1) / (n + 1)  # empirical CDF
        arg = 2 ** j0 * fn_yi - k
        # Use Haar scaling function φ(x) = I(0 ≤ x < 1)
        if 0 <= arg < 1:
            coefficients[i] = 1.0

    return np.mean(coefficients) * scale


def _local_linear_smooth(x0, grid, values, h):
    """
    Local linear smoother with Gaussian kernel at point x0.
    """
    z = grid - x0
    w = np.exp(-0.5 * (z / h) ** 2)
    w_sum = w.sum()
    if w_sum < 1e-10:
        return 0.0

    # Local linear: minimize ∑ w_i (v_i - a - b*(g_i - x0))^2
    s0 = w.sum()
    s1 = (w * z).sum()
    s2 = (w * z ** 2).sum()
    t0 = (w * values).sum()
    t1 = (w * z * values).sum()

    det = s0 * s2 - s1 ** 2
    if abs(det) < 1e-15:
        return t0 / s0 if s0 > 0 else 0.0

    a = (s2 * t0 - s1 * t1) / det
    return a


def wavelet_quantile_density(y, j0=5, bandwidth=0.15,
                              wavelet="coif1", gld_params=None):
    """
    Wavelet-based nonparametric quantile density estimation.

    Parameters
    ----------
    y : array-like
        Random sample.
    j0 : int
        Coarsest wavelet decomposition level.
    bandwidth : float
        Bandwidth for local linear smoothing.
    wavelet : str
        pywt wavelet name (default 'coif1').
    gld_params : tuple or None
        If provided (λ1, λ2, λ3, λ4), computes true GLD quantile density.

    Returns
    -------
    QuantileDensityResult
    """
    y = np.asarray(y, dtype=np.float64)
    n = len(y)

    # Step 1: Compute linear wavelet estimate
    n_basis = 2 ** j0
    a_hat = np.zeros(n_basis)
    for k in range(n_basis):
        a_hat[k] = _simpson_coef(j0, k, y)

    linear_est = 2 ** (j0 / 2) * a_hat

    # Grid
    grid = np.linspace(0, 1, n_basis)

    # Step 2: Hard thresholding via DWT
    wv = pywt.Wavelet(wavelet)
    # Use SWT for thresholding
    max_lv = min(3, pywt.swt_max_level(n_basis))
    if max_lv > 0 and n_basis >= 2 ** max_lv:
        coeffs = pywt.wavedec(linear_est, wv, level=max_lv)
        # Universal threshold
        detail_coeffs = np.concatenate(coeffs[1:])
        sigma = 1.4826 * np.median(np.abs(detail_coeffs - np.median(detail_coeffs)))
        threshold = sigma * np.sqrt(np.log(n) / n) * 8

        # Hard threshold details
        for i in range(1, len(coeffs)):
            coeffs[i] = np.where(np.abs(coeffs[i]) > threshold, coeffs[i], 0.0)

        thresholded_est = pywt.waverec(coeffs, wv)[:n_basis]
    else:
        thresholded_est = linear_est.copy()

    # Step 3: Local linear smoothing
    p_grid = np.linspace(1 / (n + 1), n / (n + 1), n)
    smoothed = np.zeros(n)
    for i in range(n):
        smoothed[i] = _local_linear_smooth(p_grid[i], grid, linear_est, bandwidth)

    # True quantile density for GLD if params given
    true_qd = None
    if gld_params is not None:
        l1, l2, l3, l4 = gld_params
        true_qd = (l3 * p_grid ** (l3 - 1) + l4 * (1 - p_grid) ** (l4 - 1)) / l2

    return QuantileDensityResult(
        grid=p_grid,
        true_qd=true_qd,
        linear_estimate=np.interp(p_grid, grid, linear_est),
        thresholded_estimate=np.interp(p_grid, grid, thresholded_est),
        smoothed_estimate=smoothed,
        sample_size=n,
        j0=j0,
        bandwidth=bandwidth,
    )

File: R/quantile_density.R
#' @title Wavelet Quantile Density Estimation
#'
#' @description Nonparametric estimator of the quantile density
#'   \eqn{q(p) = 1/f(F^{-1}(p))} using Haar scaling coefficients on the
#'   empirical CDF, with three post-processings:
#'   \itemize{
#'     \item Linear wavelet estimate;
#'     \item Hard-thresholded version via the \code{waveslim::dwt} filter;
#'     \item Local linear smoothing.
#'   }
#'
#' @param y Numeric sample.
#' @param j0 Integer. Coarsest decomposition level (\eqn{2^{j_0}} basis
#'   functions). Default 5.
#' @param bandwidth Numeric. Gaussian-kernel bandwidth for local linear
#'   smoothing.
#' @param wavelet Character. Wavelet filter for thresholding step.
#' @param gld_params Numeric length-4 \code{c(l1, l2, l3, l4)} or
#'   \code{NULL}. If provided, the true GLD quantile density is
#'   computed for ISE diagnostics.
#'
#' @return Object of class \code{"quantile_density_result"}.
#'
#' @examples
#' set.seed(1)
#' y <- rnorm(256)
#' qd <- wavelet_quantile_density(y, j0 = 4)
#' print(qd)
#'
#' @export
#' @importFrom stats sd
wavelet_quantile_density <- function(y, j0 = 5, bandwidth = 0.15,
                                     wavelet = "haar",
                                     gld_params = NULL) {
  y <- as.numeric(y)
  n <- length(y)
  n_basis <- 2L ^ j0

  # Step 1: linear wavelet (Haar scaling) coefficients via empirical CDF
  fn <- (seq_len(n)) / (n + 1)
  scale <- 2 ^ (j0 / 2)
  a_hat <- numeric(n_basis)
  for (k in 0:(n_basis - 1L)) {
    arg <- 2^j0 * fn - k
    a_hat[k + 1L] <- mean(arg >= 0 & arg < 1) * scale
  }
  linear_grid <- seq(0, 1, length.out = n_basis)
  linear_est <- scale * a_hat

  # Step 2: hard-threshold details via DWT
  thr_est <- linear_est
  trimmed <- linear_est[seq_len(2L ^ floor(log2(n_basis)))]
  if (length(trimmed) >= 4) {
    coeffs <- tryCatch(
      waveslim::dwt(trimmed, wf = wavelet,
                    n.levels = min(3L, as.integer(log2(length(trimmed)))),
                    boundary = "periodic"),
      error = function(e) NULL)
    if (!is.null(coeffs)) {
      details <- unlist(coeffs[grep("^d", names(coeffs))])
      sigma <- 1.4826 * stats::median(abs(details - stats::median(details)))
      thr <- sigma * sqrt(log(n) / n) * 8
      for (nm in grep("^d", names(coeffs), value = TRUE))
        coeffs[[nm]] <- ifelse(abs(coeffs[[nm]]) > thr, coeffs[[nm]], 0)
      thr_est_trim <- waveslim::idwt(coeffs)
      thr_est[seq_along(thr_est_trim)] <- thr_est_trim
    }
  }

  # Step 3: local linear smoothing onto a fine probability grid
  p_grid <- seq(1 / (n + 1), n / (n + 1), length.out = n)
  smoothed <- numeric(n)
  for (i in seq_len(n)) {
    z <- linear_grid - p_grid[i]
    w <- exp(-0.5 * (z / bandwidth) ^ 2)
    s0 <- sum(w); s1 <- sum(w * z); s2 <- sum(w * z^2)
    t0 <- sum(w * linear_est); t1 <- sum(w * z * linear_est)
    det <- s0 * s2 - s1 ^ 2
    smoothed[i] <- if (abs(det) < 1e-15)
      (if (s0 > 0) t0 / s0 else 0)
      else (s2 * t0 - s1 * t1) / det
  }

  true_qd <- NULL
  if (!is.null(gld_params) && length(gld_params) == 4L) {
    l <- gld_params
    true_qd <- (l[3] * p_grid ^ (l[3] - 1) +
                l[4] * (1 - p_grid) ^ (l[4] - 1)) / l[2]
  }

  res <- list(grid = p_grid, true_qd = true_qd,
              linear_estimate    = stats::approx(linear_grid, linear_est,
                                                 xout = p_grid)$y,
              thresholded_estimate = stats::approx(linear_grid, thr_est,
                                                   xout = p_grid)$y,
              smoothed_estimate = smoothed,
              sample_size = n, j0 = j0, bandwidth = bandwidth,
              call = match.call(),
              method = "Wavelet Quantile Density Estimation")
  class(res) <- "quantile_density_result"
  res
}


#' @export
print.quantile_density_result <- function(x, ...) {
  cat("\nWavelet Quantile Density Estimation\n")
  cat(strrep("=", 48), "\n", sep = "")
  cat("  N =", x$sample_size, "  coarsest j0 =", x$j0,
      "  bandwidth =", round(x$bandwidth, 3), "\n")
  if (!is.null(x$true_qd)) {
    ise <- function(a, b) mean((a - b)^2, na.rm = TRUE)
    cat(sprintf("  ISE  linear      = %.6f\n",
                ise(x$true_qd, x$linear_estimate)))
    cat(sprintf("       thresholded = %.6f\n",
                ise(x$true_qd, x$thresholded_estimate)))
    cat(sprintf("       smoothed    = %.6f\n",
                ise(x$true_qd, x$smoothed_estimate)))
  }
  invisible(x)
}

#' @export
summary.quantile_density_result <- function(object, ...) { print(object); invisible(object) }

#' @export
plot.quantile_density_result <- function(x, ...) plot_quantile_density(x, ...)
